packages/grafana-ui/src/components/Toggletip/Toggletip.tsx

```diff
@@ -0,0 +1,287 @@
+import { css } from '@emotion/css';
+import React, { useEffect, useRef } from 'react';
+import { usePopperTooltip } from 'react-popper-tooltip';
+
+import { colorManipulator, GrafanaTheme2 } from '@grafana/data';
+
+import { useStyles2 } from '../../themes/ThemeContext';
+import { IconButton } from '../IconButton/IconButton';
+import { Portal } from '../Portal/Portal';
+
+import { ToggletipContent, ToggletipPlacement } from './types';
+
+export interface ToggletipProps {
+  /** The theme used to display the toggletip */
+  theme?: 'info' | 'error';
+  /** The title to be displayed on the header */
+  title?: JSX.Element | string;
+  /** determine whether to show or not the close button **/
+  closeButton?: boolean;
+  /** Callback function to be called when the toggletip is closed */
+  onClose?: Function;
+  /** The preferred placement of the tooltip */
+  placement?: ToggletipPlacement;
+  /** The text or component that houses the content of the toggleltip */
+  content: ToggletipContent;
+  /** The text or component to be displayed on the toggletip's bottom */
+  footer?: JSX.Element | string;
+  /** The UI control users interact with to display toggletips */
+  children: JSX.Element;
+}
+
+export const Toggletip = React.memo(
+  ({
+    children,
+    theme = 'info',
+    placement = 'auto',
+    content,
+    title,
+    closeButton = true,
+    onClose,
+    footer,
+  }: ToggletipProps) => {
+    const styles = useStyles2(getStyles);
+    const style = styles[theme];
+    const contentRef = useRef(null);
+    const [controlledVisible, setControlledVisible] = React.useState(false);
+
+    useEffect(() => {
+      if (controlledVisible) {
+        const handleKeyDown = (enterKey: KeyboardEvent) => {
+          if (enterKey.key === 'Escape') {
+            setControlledVisible(false);
+            if (!!onClose) {
+              onClose();
```

Comment:
This scenario will call `onClose` twice, once here and then again when the `controlledVisible` updates to `false` and runs the `useEffect` again.
Plus this `useEffect` will run the first time this `Toggletip` is rendered and if there is an `onClose` function it will call it because `controlledVisible` is `false`. I think if we do need an `onClose` function we should not base it on the state change and actually call it any time it closes, so create a function like
```
    const closeToggletip = () => {
      setControlledVisible(false);
      onClose?.();
    };
```
and then call that here and when the close button is clicked, plus on line 74 we would also need to add a check for if it is changing to `false` call `onClose?.()`.
Then we are being more clear on when we call this `onClose` function instead of leaving it depending on the state.

Reply:
resolved on last commit